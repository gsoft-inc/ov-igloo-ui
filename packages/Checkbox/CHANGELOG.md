# @igloo-ui/checkbox

## 0.1.7

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.1.6

### Patch Changes

- 0640409: udpate classnames to 2.3.2

## 0.1.5

### Patch Changes

- d971143: All tests were updated to use testing library and not enzyme

## 0.1.4

### Patch Changes

- 0a86f13: Update content of README

## 0.1.3

### Patch Changes

- cb767c6: fix controlled state, position of checkbox with long content and the size in small space

## 0.1.2

### Patch Changes

- 8c8686e: Change @import of font.css file for a @use in SASS files

## 0.1.1

### Patch Changes

- 4a256e3: IE11 compatibility

## 0.1.0

### Minor Changes

- a903b97: Initial release

  Component API

  | name          | description                                                        |
  | ------------- | ------------------------------------------------------------------ |
  | children      | The content to display inside the label                            |
  | className     | Add a specific class to the checkbox                               |
  | dataTest      | Add a data-test tag for automated tests                            |
  | htmlFor       | Indicates the ID of the element that is controlled by the checkbox |
  | checked       | Modifies true/false value of the native checkbox                   |
  | disabled      | Modifies the native disabled state of the native checkbox          |
  | indeterminate | Applies an indeterminate state to the checkbox                     |
  | onChange      | Function called when the value changes.                            |
