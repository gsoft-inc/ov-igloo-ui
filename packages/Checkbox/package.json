{
  "name": "@igloo-ui/checkbox",
  "version": "0.1.4",
  "main": "dist/Checkbox.js",
  "module": "dist/Checkbox.js",
  "types": "dist/Checkbox.d.ts",
  "styles": "dist/checkbox.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Franck franck.gaudin@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "classnames": "^2.3.1"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
