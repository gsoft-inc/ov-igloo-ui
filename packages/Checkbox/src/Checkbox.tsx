import * as React from 'react';
import cx from 'classnames';

import './checkbox.scss';

export interface CheckboxProps extends React.ComponentPropsWithRef<'input'> {
  /** The content to display inside the label */
  children?: React.ReactNode;
  /** Add a specific class to the checkbox */
  className?: string;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Indicates the ID of the element that is controlled by the checkbox */
  htmlFor: string;
  /** Modifies true/false value of the native checkbox */
  checked?: boolean;
  /** Modifies the native disabled state of the native checkbox */
  disabled?: boolean;
  /** Applies an indeterminate state to the checkbox */
  indeterminate?: boolean;
  /** Function called when the value changes. */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Checkbox: React.FunctionComponent<CheckboxProps> = React.forwardRef(
  (props: CheckboxProps, ref: React.Ref<HTMLInputElement | null>) => {
    const {
      children,
      className,
      dataTest,
      htmlFor,
      onChange,
      checked = false,
      disabled = false,
      indeterminate = false,
      ...rest
    } = props;

    const [status, setStatus] = React.useState(checked);

    const checkRef = React.useRef<HTMLInputElement>(null);
    React.useImperativeHandle(ref, () => checkRef.current);

    React.useEffect(() => {
      setStatus(checked);
    }, [checked]);

    React.useEffect(() => {
      if (checkRef.current && indeterminate) {
        checkRef.current.indeterminate = true;
      }
    }, [indeterminate]);

    const handleOnChange = (
      event: React.ChangeEvent<HTMLInputElement>
    ): void => {
      if (onChange) {
        onChange(event);
        return;
      }

      event.stopPropagation();
      setStatus(!status);
    };

    return (
      <span className={cx('ids-form-control', className)}>
        <input
          ref={checkRef}
          id={htmlFor}
          className="ids-checkbox"
          data-test={dataTest}
          checked={status}
          disabled={disabled}
          type="checkbox"
          onChange={handleOnChange}
          {...rest}
        />
        {children && (
          <label className="ids-checkbox__label" htmlFor={htmlFor}>
            {children}
          </label>
        )}
      </span>
    );
  }
);
export default Checkbox;
