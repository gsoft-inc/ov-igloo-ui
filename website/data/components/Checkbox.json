[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Checkbox/src/Checkbox.tsx",
      "description": "",
      "displayName": "Checkbox",
      "methods": [],
      "props": {
        "children": {
          "defaultValue": null,
          "description": "The content to display inside the label",
          "name": "children",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            },
            {
              "fileName": "components/node_modules/@types/react/index.d.ts",
              "name": "TypeLiteral"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "className": {
          "defaultValue": null,
          "description": "Add a specific class to the checkbox",
          "name": "className",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "htmlFor": {
          "defaultValue": null,
          "description": "Indicates the ID of the element that is controlled by the checkbox",
          "name": "htmlFor",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": true,
          "type": { "name": "string" }
        },
        "checked": {
          "defaultValue": null,
          "description": "Modifies true/false value of the native checkbox",
          "name": "checked",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "disabled": {
          "defaultValue": null,
          "description": "Modifies the native disabled state of the native checkbox",
          "name": "disabled",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "indeterminate": {
          "defaultValue": null,
          "description": "Applies an indeterminate state to the checkbox",
          "name": "indeterminate",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "onChange": {
          "defaultValue": null,
          "description": "Function called when the value changes.",
          "name": "onChange",
          "parent": {
            "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
            "name": "CheckboxProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Checkbox/src/Checkbox.tsx",
              "name": "CheckboxProps"
            }
          ],
          "required": false,
          "type": { "name": "(e: ChangeEvent<HTMLInputElement>) => void" }
        }
      }
    }
  ]
]
