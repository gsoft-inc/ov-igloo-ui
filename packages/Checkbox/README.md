# Checkbox

Checkboxes are most commonly used to give users a way to make a range of selections (zero, one, or multiple). They may also be used as a way to have users indicate they agree to specific terms and services.

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/checkbox` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/checkbox
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/checkbox
```

## Usage

Then to use the component in your code just import it!

```jsx
import Checkbox from '@igloo-ui/checkbox';

<Checkbox htmlFor="checkbox-id">Label</Checkbox>;
```
