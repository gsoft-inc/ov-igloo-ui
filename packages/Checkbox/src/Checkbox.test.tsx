/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';

import Checkbox from './Checkbox';

const setUp = (props = {}) => {
  const component = shallow(<Checkbox htmlFor="checkbox-master" {...props} />);
  return component;
};

describe('Checkbox', () => {
  let component: any;
  beforeEach(() => {
    component = setUp();
  });

  test('It should render without errors', () => {
    const { getByText } = render(
      <Checkbox htmlFor="checkbox-test">Label</Checkbox>
    );
    expect(getByText('Label')).toBeInTheDocument();
  });

  test('It should render without label', () => {
    const { getByRole } = render(<Checkbox htmlFor="checkbox-test" />);
    expect(getByRole('checkbox')).toBeInTheDocument();
  });

  test('It should render a checked state', () => {
    const checkbox = setUp({ checked: true });
    expect(checkbox).toMatchSnapshot();
  });

  test('It should render a disabled state', () => {
    const checkbox = setUp({ disabled: true });
    expect(checkbox).toMatchSnapshot();
  });

  test('It should render a indeterminate state', () => {
    const checkbox = setUp({ indeterminate: true });
    expect(checkbox).toMatchSnapshot();
  });
});
