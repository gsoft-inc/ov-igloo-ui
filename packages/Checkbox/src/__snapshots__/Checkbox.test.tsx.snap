// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox It should render a checked state 1`] = `
<DocumentFragment>
  <span
    class="ids-form-control ids-form-control-checkbox ids-form-control-checkbox--checked"
  >
    <input
      checked=""
      class="ids-checkbox"
      id="checkbox-master"
      type="checkbox"
    />
    <label
      class="ids-checkbox__label"
      for="checkbox-master"
    >
      <span
        class="ids-checkbox__box"
      >
        <svg
          class="ids-icon ids-checkbox__check"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3.333 7.833 3.334 3.334 6-6.334"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </label>
  </span>
</DocumentFragment>
`;

exports[`Checkbox It should render a disabled state 1`] = `
<DocumentFragment>
  <span
    class="ids-form-control ids-form-control-checkbox ids-form-control-checkbox--disabled"
  >
    <input
      class="ids-checkbox"
      disabled=""
      id="checkbox-master"
      type="checkbox"
    />
    <label
      class="ids-checkbox__label"
      for="checkbox-master"
    >
      <span
        class="ids-checkbox__box"
      >
        <svg
          class="ids-icon ids-checkbox__check"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3.333 7.833 3.334 3.334 6-6.334"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </label>
  </span>
</DocumentFragment>
`;

exports[`Checkbox It should render a indeterminate state 1`] = `
<DocumentFragment>
  <span
    class="ids-form-control ids-form-control-checkbox"
  >
    <input
      class="ids-checkbox"
      id="checkbox-master"
      type="checkbox"
    />
    <label
      class="ids-checkbox__label"
      for="checkbox-master"
    >
      <span
        class="ids-checkbox__box"
      >
        <svg
          class="ids-icon ids-checkbox__check"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 8h8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </label>
  </span>
</DocumentFragment>
`;
